File: prisma/refs/java/reverse-nodes-in-k-group.java
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(0, head);
        ListNode groupPrev = dummy;
        for (;;) {
            ListNode kth = groupPrev;
            for (int i = 0; i < k && kth != null; i++) kth = kth.next;
            if (kth == null) break;
            ListNode groupNext = kth.next;
            ListNode prev = groupNext, cur = groupPrev.next;
            while (cur != groupNext) {
                ListNode nx = cur.next;
                cur.next = prev;
                prev = cur;
                cur = nx;
            }
            ListNode newStart = groupPrev.next;
            groupPrev.next = kth;
            groupPrev = newStart;
        }
        return dummy.next;
    }
}

File: prisma/refs/py/reverse-nodes-in-k-group.py
def reverse_k_group(head, k):
    dummy = ListNode(0, head)  # noqa: F821 (injected by harness)
    group_prev = dummy
    while True:
        kth = group_prev
        for _ in range(k):
            kth = kth.next
            if kth is None:
                return dummy.next
        group_next = kth.next
        prev, cur = group_next, group_prev.next
        while cur is not group_next:
            cur.next, prev, cur = prev, cur, cur.next
        new_start = group_prev.next
        group_prev.next = kth
        group_prev = new_start
